File: Classes/GameScene.cpp
//
//  GameScene.cpp
//  FlappyBird
//
//  Created by 夏旸 on 16/7/19.
//
//

#include "GameScene.h"
#include "BackgroundLayer.h"
#include "OptionLayer.h"
#include "AtlasLoader.h"
#include "SimpleAudioEngine.h"
#include "StatusLayer.h"
#include "UserRecord.h"

using namespace CocosDenshion;

Scene *GameScene::createScene()
{
    auto scene = Scene::createWithPhysics();
    scene->getPhysicsWorld()->setGravity(Vect(0, -900));
	//背景
    auto backgroundLayer = BackgroundLayer::create();
    //状态
	auto statusLayer = StatusLayer::create();
	//游戏
    auto gameLayer = GameScene::create();
	gameLayer->setDelegator(statusLayer);
	gameLayer->gameScene = scene;

	// 触摸事件
	auto optionLayer = OptionLayer::create();
	optionLayer->setDelegator(gameLayer);
	
	scene->addChild(backgroundLayer);

	scene->addChild(gameLayer);
    
	scene->addChild(statusLayer);

	scene->addChild(optionLayer);

    return scene;
}

GameScene::GameScene(){}
GameScene::~GameScene(){}

bool GameScene::init()
{
    if (!Layer::init()) {
        return false;
    }
    Size visiableSize = Director::getInstance()->getVisibleSize();
    Point origin = Director::getInstance()->getVisibleOrigin();
    
    this->gameStatus = GAME_STATUS_READY;
    this->score = 0;
    
	this->bird = BirdSprite::getInstance()->createBird();
    this->bird->setPosition(origin.x + visiableSize.width*1/3-5, origin.y + visiableSize.height/2+5);
	BirdSprite::getInstance()->idle(this->bird);
    this->addChild(this->bird);
    
	// Add ground
    auto groundNode = Node::create();
    float landHeight = BackgroundLayer::getLandHeight();
    auto groundBody = PhysicsBody::create();
    groundBody->addShape(PhysicsShapeBox::create(Size(288, landHeight)));
    groundBody->setDynamic(false);
    groundBody->setLinearDamping(0.0f);
    groundBody->setContactTestBitmask(0xFFFFFFFF);
    groundNode->setPhysicsBody(groundBody);
    groundNode->setPosition(144, landHeight/2);
    addChild(groundNode);
    
    this->landSprite1 = Sprite::createWithSpriteFrame(AtlasLoader::getInstance()->getSpriteFrameByName("land"));
    this->landSprite1->setAnchorPoint(Point::ZERO);
    this->landSprite1->setPosition(Point::ZERO);
    this->addChild(this->landSprite1, 30);

    this->landSprite2 = Sprite::createWithSpriteFrame(AtlasLoader::getInstance()->getSpriteFrameByName("land"));
    this->landSprite2->setAnchorPoint(Point::ZERO);
    this->landSprite2->setPosition(this->landSprite1->getContentSize().width-2.0f,0);
    this->addChild(this->landSprite2, 30);
    
    this->shiftLand = schedule_selector(GameScene::scrollLand);
    this->schedule(this->shiftLand, 0.01f);
    this->scheduleUpdate();

    return true;
}

void GameScene::restartGame()
{
	BirdSprite::getInstance() ->die(this->bird);
    this->removeChild(this->bird);

	auto scene = GameScene::createScene();
	TransitionScene *transition = TransitionFade::create(1, scene);
	Director::getInstance()->replaceScene(transition);
}

bool GameScene::onContactBegin(const PhysicsContact& contact) {
    this->bird->getPhysicsBody()->setVelocity(Vect::ZERO);
    this->bird->getPhysicsBody()->setGravityEnable(false);
    this->gameOver();
    return true;
}

void GameScene::scrollLand(float dt){
    this->landSprite1->setPositionX(this->landSprite1->getPositionX() - 2.0f);
    this->landSprite2->setPositionX(this->landSprite1->getPositionX() + this->landSprite1->getContentSize().width - 2.0f);
    if(this->landSprite2->getPositionX() == 0) {
        this->landSprite1->setPositionX(0);
    }
    
    // move the pips
    for (auto singlePip : this->pips) {
        singlePip->setPositionX(singlePip->getPositionX() - 2);
        if(singlePip->getPositionX() < -PIP_WIDTH) {
            singlePip->setTag(PIP_NEW);
            Size visibleSize = Director::getInstance()->getVisibleSize();
            singlePip->setPositionX(visibleSize.width);
            singlePip->setPositionY(this->getRandomHeight());
        }
    }
}

void GameScene::onTouch() {
    if(this->gameStatus == GAME_STATUS_OVER) {
        return;
    }
    SimpleAudioEngine::getInstance()->playEffect("sfx_wing.ogg");
    if(this->gameStatus == GAME_STATUS_READY) {
		PhysicsBody *body = PhysicsBody::create();
		body->addShape(PhysicsShapeCircle::create(BIRD_RADIUS));
		body->setLinearDamping(.0f);
		body->setContactTestBitmask(0xFFFFFFFF);
		this->bird->setPhysicsBody(body);
        this->delegator->onGameStart();
		BirdSprite::getInstance()->fly(this->bird);
        this->gameStatus = GAME_STATUS_START;
        this->createPips();
    }
    this->bird->getPhysicsBody()->setVelocity(Vect(0, 260));
    this->rotateBird();
}

void GameScene::rotateBird() {
    float verticalSpeed = this->bird->getPhysicsBody()->getVelocity().y;
    this->bird->setRotation(-MIN(MAX(-90, (verticalSpeed*0.2 + 60)), 30));
}


void GameScene::update(float delta) {
    if (this->gameStatus == GAME_STATUS_START) {
        this->rotateBird();
        this->checkHit();
    }
}

void GameScene::onEnter()
{
    Layer::onEnter();
    auto contactListener = EventListenerPhysicsContact::create();
    contactListener->onContactBegin = CC_CALLBACK_1(GameScene::onContactBegin, this);
    Director::getInstance()->getEventDispatcher()->addEventListenerWithSceneGraphPriority(contactListener, this);
}

void GameScene::onExit()
{
    Layer::onExit();
    Director::getInstance()->getEventDispatcher()->removeEventListenersForTarget(this);
}

void GameScene::createPips() {
    // Create the pips
    for (int i = 0; i < PIP_COUNT; i++) {
        Size visibleSize = Director::getInstance()->getVisibleSize();
        Sprite *pipUp = Sprite::createWithSpriteFrame(AtlasLoader::getInstance()->getSpriteFrameByName("pipe_up"));
        Sprite *pipDown = Sprite::createWithSpriteFrame(AtlasLoader::getInstance()->getSpriteFrameByName("pipe_down"));
        Node *singlePip = Node::create();
        
        // bind to pair
        pipDown->setPosition(0, PIP_HEIGHT + PIP_DISTANCE);
        singlePip->addChild(pipDown, 0, DOWN_PIP);
        singlePip->addChild(pipUp, 0, UP_PIP);
        singlePip->setPosition(visibleSize.width + i*PIP_INTERVAL + WAIT_DISTANCE, this->getRandomHeight());
        auto body = PhysicsBody::create();
        auto shapeBoxDown = PhysicsShapeBox::create(pipDown->getContentSize(),PHYSICSSHAPE_MATERIAL_DEFAULT, Point(0, PIP_HEIGHT + PIP_DISTANCE));
        body->addShape(shapeBoxDown);
        body->addShape(PhysicsShapeBox::create(pipUp->getContentSize()));
        body->setDynamic(false);
        body->setContactTestBitmask(0xFFFFFFFF);
        singlePip->setPhysicsBody(body);
        singlePip->setTag(PIP_NEW);
        
        this->addChild(singlePip);
        this->pips.push_back(singlePip);
    }
}

int GameScene::getRandomHeight() {
    Size visibleSize = Director::getInstance()->getVisibleSize();
    return rand()%(int)(2*PIP_HEIGHT + PIP_DISTANCE - visibleSize.height);
}

void GameScene::checkHit() {
    for(auto pip : this->pips) {
        if (pip->getTag() == PIP_NEW) {
            if (pip->getPositionX() < this->bird->getPositionX()) {
                SimpleAudioEngine::getInstance()->playEffect("sfx_point.ogg");
                this->score ++;
                this->delegator->onGamePlaying(this->score);
                pip->setTag(PIP_PASS);
            }
        }
    }
}

void GameScene::gameOver() {
    if(this->gameStatus == GAME_STATUS_OVER) {
        return;
    }
    SimpleAudioEngine::getInstance()->playEffect("sfx_hit.ogg");
    
    int bestScore = UserRecord::getInstance()->readIntegerFromUserDefault("best_score");
    //update the best score
    if(this->score > bestScore){
        UserRecord::getInstance()->saveIntegerToUserDefault("best_score",this->score);
    }
    this->delegator->onGameEnd(this->score, bestScore);
    
    this->unschedule(shiftLand);
    SimpleAudioEngine::getInstance()->playEffect("sfx_die.ogg");
	BirdSprite::getInstance()->die(this->bird);
    this->bird->setRotation(90);
    this->birdSpriteFadeOut();
    this->gameStatus = GAME_STATUS_OVER;
}

void GameScene::birdSpriteFadeOut(){
    FadeOut* animation = FadeOut::create(1.5);
    CallFunc* animationDone = CallFunc::create(bind(&GameScene::birdSpriteRemove,this));
    Sequence* sequence = Sequence::createWithTwoActions(animation,animationDone);
    this->bird->stopAllActions();
    this->bird->runAction(sequence);
}

void GameScene::birdSpriteRemove(){
    this->bird->setRotation(0);
    this->removeChild(this->bird);
}

File: src/app/views/GameScene.lua

cc.exports.GameStatus = 
{
	GAME_STATUS_READY = 0,
	GAME_STATUS_START = 1,
	GAME_STATUS_OVER = 2
}

cc.exports.UP_PIP = 21
cc.exports.DOWN_PIP = 12
cc.exports.PIP_PASS = 30
cc.exports.PIP_NEW = 31
cc.exports.BIRD_RADIUS = 15
cc.exports.PIP_HEIGHT = 320
cc.exports.PIP_WIDTH = 52
cc.exports.PIP_SHIFT_SPEED = 80
cc.exports.PIP_DISTANCE = 120
cc.exports.PIP_INTERVAL = 180
cc.exports.PIP_COUNT = 2
cc.exports.WAIT_DISTANCE = 100

local oggWing = "sfx_wing.ogg"
local oggPoint = "sfx_point.ogg"
local oggHit = "sfx_hit.ogg"
local oggDie = "sfx_die.ogg"
local pipeUpName = "pipe_up"
local pipeDownName = "pipe_down"
local landName = "land"
local saveScoreKey = "best_score"

local GameScene = class("GameScene", cc.Layer)

local birdSprite = require("app.tools.BirdSprite"):getInstance()
local atlasLoader = require("app.tools.AtlasLoader"):getInstance()
local director = cc.Director:getInstance()
local visibleSize = director:getVisibleSize()
local origin = director:getVisibleOrigin()

function GameScene:createScene()
	local scene = cc.Scene:createWithPhysics()
	scene:getPhysicsWorld():setGravity({ x=0, y=-900 })
	-- scene:getPhysicsWorld():setDebugDrawMask(cc.PhysicsWorld.DEBUGDRAW_ALL)
	local backgroundLayer = require("app.views.BackgroundLayer"):create()
	local statusLayer = require("app.views.StatusLayer"):create()
	local gameLayer = GameScene:create()
	local optionLayer = require("app.views.OptionLayer"):create()

	gameLayer.statusDelegate = statusLayer
	gameLayer.gameScene = scene

	optionLayer.gameDelegate = gameLayer

	scene:addChild(backgroundLayer)
	scene:addChild(gameLayer)
	scene:addChild(statusLayer)
	scene:addChild(optionLayer)
	return scene
end

function GameScene:ctor()
	math.randomseed(os.time())
	
	self.gameStatus = GameStatus.GAME_STATUS_READY
	self.score = 0
	self.pips = {}

	--land1
	self.land1 = cc.Sprite:createWithSpriteFrame(atlasLoader:getSpriteFrameByName(landName))
	self.land1:setAnchorPoint({ x = 0, y = 0 })
	self.land1:setPosition({ x = 0, y = 0 })
	self:addChild(self.land1)
	--land2
	self.land2 = cc.Sprite:createWithSpriteFrame(atlasLoader:getSpriteFrameByName(landName))
	self.land2:setAnchorPoint({ x = 0, y = 0 })
	self.land2:setPosition({ x = self.land1:getContentSize().width - 2.0, y = 0 })
	self:addChild(self.land2)

	self.groundNode = cc.Node:create()
	local landHeight = self.land1:getContentSize().height
	local groundBody = cc.PhysicsBody:create()
	groundBody:addShape(cc.PhysicsShapeBox:create({ width=288, landHeight }))
	groundBody:setDynamic(false)
	groundBody:setLinearDamping(0.0)
	groundBody:setContactTestBitmask(0xFFFFFFFF)
	self.groundNode:setPhysicsBody(groundBody)
	self.groundNode:setPosition(144, landHeight)
	self:addChild(self.groundNode)

	self.bird = birdSprite:createBird()
	self.bird:setPosition({ x=origin.x + visibleSize.width/3-5, y=origin.y + visibleSize.height/2+5 })
	birdSprite:idle(self.bird)
	self:addChild(self.bird)

	--schedule
	self.scrollLandSchedule = director:getScheduler():scheduleScriptFunc(handler(self, self.scrollHand), 0.01, false)
	self:scheduleUpdateWithPriorityLua(handler(self, self.update), 0)
	--onEnter onExit 事件分发
	self:registerScriptHandler(handler(self, self.onNodeEvent))
end

function GameScene:onNodeEvent(event)
	if event == "enter" then
		self:onEnter()
	elseif event == "exit" then
		self:onExit()
	end
end

function GameScene:restartGame()
	birdSprite:die(self.bird)
	self:removeChild(self.bird)
	self.bird = nil

	local scene = GameScene:createScene()
	local transition = cc.TransitionFade:create(1.3, scene)
	cc.Director:getInstance():replaceScene(transition)
end

function GameScene:onContactBegin(contact)
	self.bird:getPhysicsBody():setVelocity({ x=0, y=0 })
	self.bird:getPhysicsBody():setGravityEnable(false)
	self:gameOver()
end

function GameScene:update(dt)
	if(self.gameStatus == GameStatus.GAME_STATUS_START) then
		self:rotateBird()
		self:checkHit()
	end
end

function GameScene:scrollHand(dt)
	self.land1:setPositionX(self.land1:getPositionX() - 2.0)
	self.land2:setPositionX(self.land1:getPositionX() + self.land1:getContentSize().width - 2.0)
	if(self.land2:getPositionX() == 0) then
		self.land1:setPositionX(0)
	end

	for k in ipairs(self.pips) do
		self.pips[k]:setPositionX(self.pips[k]:getPositionX() - 2)
		if ((self.pips[k]:getPositionX()) < (-PIP_WIDTH)) then
			self.pips[k]:setTag(PIP_NEW)
			self.pips[k]:setPositionX(visibleSize.width)
			self.pips[k]:setPositionY(self:getRandomHeight())
		end
	end
end 

function GameScene:onEnter()
	local contactListener = cc.EventListenerPhysicsContact:create()
	contactListener:registerScriptHandler(handler(self, self.onContactBegin), cc.Handler.EVENT_PHYSICS_CONTACT_BEGIN)
	cc.Director:getInstance():getEventDispatcher():addEventListenerWithSceneGraphPriority(contactListener, self)
end

function GameScene:onExit()
	self:unregisterScriptHandler()
	cc.Director:getInstance():getEventDispatcher():removeEventListenersForTarget(self)
end

function GameScene:onTouch()
	if self.gameStatus == GameStatus.GAME_STATUS_OVER then
		return
	end
	cc.SimpleAudioEngine:getInstance():playEffect(oggWing)
	if self.gameStatus == GameStatus.GAME_STATUS_READY then
		local body = cc.PhysicsBody:create()
		body:addShape(cc.PhysicsShapeCircle:create(BIRD_RADIUS))
		body:setLinearDamping(0.0)
		body:setContactTestBitmask(0xFFFFFFFF)
		self.bird:setPhysicsBody(body)
		if self.statusDelegate then
			self.statusDelegate:onGameStart()
		end
		birdSprite:fly(self.bird)
		self.gameStatus = GameStatus.GAME_STATUS_START
		self:createPips()
	end
	self.bird:getPhysicsBody():setVelocity({ x=0, y=260 })
	self:rotateBird()
end

function GameScene:rotateBird()
	local verticalSpeed = self.bird:getPhysicsBody():getVelocity().y
	self.bird:setRotation(-math.min(math.max(-90, verticalSpeed*0.2+60), 30))
end

function GameScene:createPips() 
    for i = 0, PIP_COUNT-1 do
        local pipUp = cc.Sprite:createWithSpriteFrame(atlasLoader:getSpriteFrameByName(pipeUpName))
        local pipDown = cc.Sprite:createWithSpriteFrame(atlasLoader:getSpriteFrameByName(pipeDownName))
        local singlePip = cc.Node:create()
        
        pipUp:setPosition({ x=0, y=0 })
        pipDown:setPosition({ x=0, y=PIP_HEIGHT + PIP_DISTANCE})
        singlePip:addChild(pipDown, 0, DOWN_PIP)
        singlePip:addChild(pipUp, 0, UP_PIP)
        singlePip:setPosition({ x=visibleSize.width + i*PIP_INTERVAL + WAIT_DISTANCE, y=self:getRandomHeight() })
        local body = cc.PhysicsBody:create()
        local shapeBoxDown = cc.PhysicsShapeBox:create(pipDown:getContentSize(), cc.PHYSICSSHAPE_MATERIAL_DEFAULT, { x=0, y=PIP_HEIGHT + PIP_DISTANCE })
        body:addShape(shapeBoxDown)
        body:addShape(cc.PhysicsShapeBox:create(pipUp:getContentSize()))
        body:setDynamic(false)
        body:setContactTestBitmask(0xFFFFFFFF)
        singlePip:setPhysicsBody(body)
        singlePip:setTag(PIP_NEW)
        
        self:addChild(singlePip)
        table.insert(self.pips, singlePip)
    end
end

function GameScene:getRandomHeight() 
    return (math.random(PIP_HEIGHT/2) - 1)
end

function GameScene:checkHit() 
    for k in ipairs(self.pips) do
        if (self.pips[k]:getTag() == PIP_NEW) then
            if (self.pips[k]:getPositionX() < self.bird:getPositionX()) then
                cc.SimpleAudioEngine:getInstance():playEffect(oggPoint)
                self.score = self.score + 1
                self.statusDelegate:onGamePlaying(self.score)
                self.pips[k]:setTag(PIP_PASS)
            end
        end
    end
end

function GameScene:gameOver() 
    if (self.gameStatus == GameStatus.GAME_STATUS_OVER) then
        return
    end
    cc.SimpleAudioEngine:getInstance():playEffect(oggHit)
    local userRecord = require("app.tools.UserRecord"):getInstance()
    local bestScore = userRecord:readIntegerFromUserDefault(saveScoreKey)
    if (self.score > bestScore)  then
        userRecord:saveIntegerToUserDefault(saveScoreKey, self.score)
    end
    if self.statusDelegate then
    	self.statusDelegate:onGameEnd(self.score, bestScore)
	end
    director:getScheduler():unscheduleScriptEntry(self.scrollLandSchedule)
    cc.SimpleAudioEngine:getInstance():playEffect(oggDie)
	birdSprite:getInstance():die(self.bird)
    self.bird:setRotation(90)
    self:birdSpriteFadeOut()
    self.gameStatus = GameStatus.GAME_STATUS_OVER;
end

function GameScene:birdSpriteFadeOut()
    local animation = cc.FadeOut:create(1.5)
    local animationDone = cc.CallFunc:create(handler(self, self.birdSpriteRemove))
    local sequence = cc.Sequence:create(animation, animationDone)
    self.bird:stopAllActions()
    self.bird:runAction(sequence)
end

function GameScene:birdSpriteRemove()
    self.bird:setRotation(0)
    self:removeChild(self.bird)
    self.bird = nil
end

return GameScene
